project_euler/problem_188/sol1.py

```diff
@@ -0,0 +1,53 @@
+"""
+The hyperexponentiation of a number
+Problem 188
+
+The hyperexponentiation or tetration of a number a by a positive integer b,
+denoted by a↑↑b or b^a, is recursively defined by:
+
+a↑↑1 = a,
+a↑↑(k+1) = a(a↑↑k).
+
+Thus we have e.g. 3↑↑2 = 3^3 = 27, hence 3↑↑3 = 3^27 = 7625597484987 and
+3↑↑4 is roughly 103.6383346400240996*10^12.
+
+Find the last 8 digits of 1777↑↑1855.
+"""
+
+
+def solution(a=1777, k=1855, digits=8):
+    """Returns the last 8 digits of the hyperexponentiation of a by k.
+
+    >>> solution()
+    95962097
+    >>> solution(3, 2)
+    27
+    >>> solution(3, 3)
+    97484987
+    """
+
+    # we calculate everything modulo 10^8, since we only care about the
+    # last 8 digits
+    modulo_value = 10 ** digits
+
+    # small helper function for modular exponentiation, to keep the result
+    # values small enough
+    def modexpt(base, exponent):
```

Comment:
You can keep this function separately.
Include type hints, `doctest` for this function